src/main/java/sirius/db/mixing/query/QueryCompiler.java

```diff
@@ -373,13 +373,15 @@ private boolean continueToken(boolean inQuotes) {
             return Tuple.create(effectiveMapping, property);
         }
 
-        if (!(property instanceof BaseEntityRefProperty)) {
-            return null;
+        // properly resolve reference properties
+        if (property instanceof BaseEntityRefProperty<?, ?, ?> baseEntityRefProperty) {
+            return resolveProperty(baseEntityRefProperty.getReferencedDescriptor(), splitPath.getSecond(), effectiveMapping);
         }
 
-        return resolveProperty(((BaseEntityRefProperty<?, ?, ?>) property).getReferencedDescriptor(),
-                               splitPath.getSecond(),
-                               effectiveMapping);
+        // reaching this point, we likely have steampunk database access with most complicated structures; just add the
```

Comment:
as far as I understand it, our most common use-case is indexing into a StringMap or into a NestedList? Couldn't we handle this case by simply returning Tuple(Mapping.named(splitPath.getSecond()), property);

And then overwrite the QueryCompiler.compileFieldEquals to also detect properties of this kind and then generate the proper constraint (as implemented above). This way, we woudln't have to change the FilterFactory at all, and we would be sort of sure, that we properly handle the query in a predictive way (or would throw an error otherwise)...

Reply:
I left the `join` as more elegant expression than `inner`; Mongo seems to handle it alright, and we have special processing for Elastic anyways, and the recursive structure occurs to match the DB structure more accurately. Correct me if I am wrong. 😅